### ISSUE TITLE
Don't force the use of FastBufferedWriter

### ISSUE THREAD

**Issue by qtxo** [osiegmar/FastCSV#7]
Many Writer implementations are already buffered and fast enough. Also, many use cases start with an already existing BufferedWriter or similar, adding an extra layer only adds another temporary buffer copy.
For example, you may have code like this working for different uses cases:
```
GZIPOutputStream zout = new GZIPOutputStream(
new CipherOutputStream(new FileOutputStream(file), c));
		
return new BufferedWriter(new OutputStreamWriter(zout, "utf-8"));
```
Assuming you can't change the code above or you don't want to be forced to do this:

```
if (usingFasctCsv)
   return new OutputStreamWriter(zout, "utf-8")
else 
   return new BufferedWriter(new OutputStreamWriter(zout, "utf-8"));
```

Can you add some way to construct an appender without wrapping the writer? 
Or I will try a pull request...

**Comment by osiegmar** [comment]
I've done heavy performance tests on that. The numbers speak for themselves:

* Current implementation (with passing a non-buffered Writer): 7,224,961 records per second
* Current implementation (with passing a BufferedWriter): 3,506,184 records per second
* Replace FastBufferedWriter with BufferedWriter: 3,378,508 records per second

So, if speed is not the reason why you picked FastCSV, passing a BufferedWriter to FastCSV is still faster than removing/replacing FastBufferedWriter completely.

BTW: The GZIP-code in your example is using the BufferedWriter in the wrong way. See [What order should I use GzipOutputStream and BufferedOutputStream](https://stackoverflow.com/questions/1082320/what-order-should-i-use-gzipoutputstream-and-bufferedoutputstream).

**Comment by qtxo** [comment]
Sorry, I think you read too fast :)
The question you linked has 0 references to BufferedWriter :) 

But if you look at [OutputStreamWriter javadoc](https://docs.oracle.com/javase/8/docs/api/java/io/OutputStreamWriter.html) it explicitly states:

```
For top efficiency, consider wrapping an OutputStreamWriter within a BufferedWriter so as to avoid frequent converter invocations. For example:

 Writer out
   = new BufferedWriter(new OutputStreamWriter(System.out));
```
So all the good citizens of Java using your (excellent) library will suffer another buffering layer.

My point is not about performance, it is about ergonomics.

**Comment by osiegmar** [comment]
My comment about the GZIP use, was just because of the non-ideal order of the buffer. The GZIPOutputStream already has a buffer. And performance tests (on the linked page) has shown, that this can only be slightly better when wrapping the FileOutput with a buffer. But this was only BTW -- not our topic here.

I don't get your point about ergonomics. Just use your BufferedWriter with FastCSV -- the performance does not suffer from the additional buffering layer -- as my test have proven.

And please don't forget, that performance is the number one goal of FastCSV as the name implies.

**Comment by qtxo** [comment]
Thanks for the comment. The order is the correct one, in fact changing the order makes no sense at all.

There are 2 types of buffers involved, the ones for the OutputStream low level stuff, and the other is to optimize the frequent invocations of StreamEncoder in OutputStreamWriter as the javadoc says, I am talking about this one.

For ergonomics, (sorry I thought I'd put it in the original question) is about the "flush()"
**Your current implementation forces us to call CsvAppender.flush() at the end, when the common and standard approach is to call flush on the Writer**. Hope I am stating the problem clearly: without calling CsvAppender.flush() at the end the last buffer will never be written. If you wrap the passed writer then you should return it so callers without internal knowledge or libraries can call writer.flush() and writer.close() only.

Anyway, it is good to know that the performance will be good, in any case for me the gzip and encryption weight much more on the whole.

**Comment by qtxo** [comment]
Can you agree that this code makes sense? This fixes the problem.

```
if (writer instanceof FastBufferedWriter)  {
            this.writer = writer;
        } else {
            this.writer = new FastBufferedWriter(writer);
        }
```
If you accept a pull request I will send it right now.

Although if you were to start from scratch (suggestion for version 2?), this constructor would avoid Writer surprises:
```
CsvAppender(final FastBufferedWriter writer, final char fieldSeparator, final char textDelimiter,...
```

**Comment by osiegmar** [comment]
Now, I got your point. Somewhat.

> Your current implementation forces us to call CsvAppender.flush() at the end, when the common and standard approach is to call flush on the Writer.

The CsvAppender is meant to **be** the Writer. I agree, that the name CsvAppender is misleading here -- maybe CsvWriter would have been a better name for it. And as it is the rule for all chains of streams/writers -- only call flush()/close() on the outermost one.

The FastBufferedWriter is meant only for internal buffering. I created it only not to clutter the code of CsvAppender with internal buffering.

I will not change this is 1.x but will reconsider it for 2.x.

Thanks for pointing that out!
